src/cpp/session/modules/SessionConnections.R

```diff
@@ -527,20 +527,27 @@ options(
    do.call(rbind, lapply(registryEntriesValue, function(e) data.frame(e, stringsAsFactors = FALSE)))
 })
 
-.rs.addFunction("connectionReadOdbcEntry", function(drivers, uniqueDriverNames, driver) {
+.rs.addFunction("connectionReadOdbcEntry", function(drivers, driver) {
    tryCatch({
-      currentDriver <- drivers[drivers$attribute == "Driver" & drivers$name == driver, ]
-      driverInstaller <- drivers[drivers$attribute == "Installer" & drivers$name == driver, ]
-      driverId <- gsub(.rs.connectionOdbcRStudioDriver(), "", driver)
-
-      basePath <- sub(paste(tolower(driver), ".*$", sep = ""), "", currentDriver$value)
-      snippetsFile <- file.path(
-         basePath,
-         tolower(driver),
-         "snippets",
-         paste(tolower(driverId), ".R", sep = "")
-      )
-      
+      currentDriver <- drivers[drivers$attribute == "Driver" & drivers$name == driver, ]$value
+      driverInstaller <- drivers[drivers$attribute == "Installer" & drivers$name == driver, ]$value
+      driverId <- .rs.connectionStripRStudioDriver(driver)
+
+      # Instead of assuming the location of the snippets directory, we instead
+      # walk up the directory tree to find the snippets directory
+      prevDir <- NULL
+      walkDir <- dirname(currentDriver)
+      while (!identical(walkDir, prevDir)) {
+         snippetsDir <- file.path(walkDir, "snippets")
+         if (dir.exists(snippetsDir)) {
+            break
+         }
+         prevDir <- walkDir
+         walkDir <- dirname(walkDir)
+      }
+
+      snippetsFile <- file.path(snippetsDir, paste(tolower(driverId), ".R", sep = ""))
```

Comment:
Do we want to consider stopping this at the home directory? 

This could also struggle when ACLs are involved since we've had issues before with another piece of (C++) code that walks up the directory structure when ACLs are involved. I don't think it's a huge issue here, though.

Reply:
Good thinking on the ACLs.

We can't just stop at the user's home directory because it also will read entries from the system ODBC config like
* `/opt/simba/spark/lib/64/libsparkodbc_sb64.so`
* `/opt/rstudio-drivers/postgresql/bin/lib/libpostgresqlodbc_sb64.so`